ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -773,30 +790,61 @@ decodeByronHeaderHash :: Decoder s (HeaderHash (ByronBlockOrEBB cfg))
 decodeByronHeaderHash = fromCBOR
 
 encodeByronGenTx :: GenTx (ByronBlockOrEBB cfg) -> Encoding
-encodeByronGenTx (ByronTx {byronTx = tx }) = toCBOR (void tx)
+encodeByronGenTx genTx = toCBOR $ void (mkMempoolPayload genTx)
```

Comment:
Unrelated to your changes, but I just noticed that `toCBOR . void` throws away the serialisation we have and recomputes it. Can't we reuse it somehow? :thinking: Problem: we have a `ByteString`, but need an `Encoding`.

Reply:
This could be achieved by adding an `annotation` field to the `AMempoolPayload` data type.

Reply:
To reuse the `ByteString`, I think ideally we would want to just annotate the `AMempoolPayload a` (add an `annotation` field to it in `cardano-ledger` or use an `Annotated MempoolPayload a`). However, our `GenTx` definition diverges a bit from that of `AMempoolPayload` as we need to additionally cache the hash/ID of each ledger object.

I guess one option would be to include the lazy hash of each ledger object within the actual `AMempoolPayload` definition in `cardano-ledger`. Then we would change the definition of the Byron `GenTx` to something like: 

```
data GenTx (ByronBlockOrEBB cfg) = ByronGenTx !(AMempoolPayload ByteString)
```

Then, if we add an `annotation :: a` field to `AMempoolPayload a`, we'd be able to do:

```
encodeByronGenTx :: GenTx (ByronBlockOrEBB cfg) -> Encoding
encodeByronGenTx (ByronGenTx mp) = annotation mp
```

Reply:
Umm, one option would be to annotate each of the `GenTx` constructors or something to that effect.

I can also think of another option but it could also look a bit ugly. Something like:

```
...
encodeByronGenTx (ByronDlg _ cert) =
  encodeListLen 2 <> toCBOR (1 :: Word8) <> encodeBytes (annotation cert)
encodeByronGenTx (ByronUpdateProposal _ proposal) =
  encodeListLen 2 <> toCBOR (2 :: Word8) <> encodeBytes (annotation proposal)
...
```

Reply:
Umm, one option would be to annotate each of the `GenTx` constructors or something to that effect, but annotating a `GenTx` constructor with a `ByteString` that represents a `MempoolPayload` seems very odd.

I can also think of another option. Something like:

```
...
encodeByronGenTx (ByronDlg _ cert) =
  encodeListLen 2 <> toCBOR (1 :: Word8) <> encodeBytes (annotation cert)
encodeByronGenTx (ByronUpdateProposal _ proposal) =
  encodeListLen 2 <> toCBOR (2 :: Word8) <> encodeBytes (annotation proposal)
...
```

This is pretty much a copy of the `MempoolPayload` `ToCBOR` instance. Only difference is the `encodeBytes` part.

Reply:
ah yes, that makes sense .I'll see how this works out.

Reply:
You solved this in https://github.com/input-output-hk/cardano-ledger/pull/660.